src/wbase/ChannelShared.cc

```diff
@@ -0,0 +1,337 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+// Class header
+#include "wbase/ChannelShared.h"
+
+// Qserv headers
+#include "global/LogContext.h"
+#include "proto/ProtoHeaderWrap.h"
+#include "qmeta/types.h"
+#include "util/Bug.h"
+#include "util/Error.h"
+#include "wcontrol/TransmitMgr.h"
+#include "wbase/Task.h"
+#include "wpublish/QueriesAndChunks.h"
+
+// LSST headers
+#include "lsst/log/Log.h"
+
+using namespace std;
+
+namespace {
+LOG_LOGGER _log = LOG_GET("lsst.qserv.wbase.ChannelShared");
+}
+
+namespace lsst::qserv::wbase {
+
+atomic<uint64_t> ChannelShared::scsSeqId{0};
+
+ChannelShared::ChannelShared(shared_ptr<wbase::SendChannel> const& sendChannel,
+                             shared_ptr<wcontrol::TransmitMgr> const& transmitMgr, qmeta::CzarId czarId)
+        : _sendChannel(sendChannel), _transmitMgr(transmitMgr), _czarId(czarId), _scsId(scsSeqId++) {
+    if (_sendChannel == nullptr) {
+        throw util::Bug(ERR_LOC, "ChannelShared constructor given nullptr");
+    }
+}
+
+ChannelShared::~ChannelShared() {
+    if (_sendChannel != nullptr) {
+        _sendChannel->setDestroying();
+        if (!_sendChannel->isDead()) {
+            _sendChannel->kill("~ChannelShared()");
+        }
+    }
+}
+
+bool ChannelShared::send(char const* buf, int bufLen) {
+    lock_guard<mutex> const streamMutexLock(_streamMutex);
+    return _sendChannel->send(buf, bufLen);
+}
+
+bool ChannelShared::sendError(string const& msg, int code) {
+    lock_guard<mutex> const streamMutexLock(_streamMutex);
+    return _sendChannel->sendError(msg, code);
+}
+
+bool ChannelShared::sendFile(int fd, wbase::SendChannel::Size fSize) {
+    lock_guard<mutex> const streamMutexLock(_streamMutex);
+    return _sendChannel->sendFile(fd, fSize);
+}
+
+bool ChannelShared::sendStream(xrdsvc::StreamBuffer::Ptr const& sBuf, bool last, int scsSeq) {
+    lock_guard<mutex> const streamMutexLock(_streamMutex);
+    return _sendChannel->sendStream(sBuf, last, scsSeq);
+}
+
+bool ChannelShared::kill(string const& note) {
+    lock_guard<mutex> const streamMutexLock(_streamMutex);
+    return _kill(streamMutexLock, note);
+}
+
+bool ChannelShared::isDead() {
+    if (_sendChannel == nullptr) return true;
+    return _sendChannel->isDead();
+}
+
+void ChannelShared::setTaskCount(int taskCount) { _taskCount = taskCount; }
+
+bool ChannelShared::transmitTaskLast() {
+    lock_guard<mutex> const streamMutexLock(_streamMutex);
+    ++_lastCount;
+    bool lastTaskDone = _lastCount >= _taskCount;
+    return lastTaskDone;
```

Comment:
This should have it's own mutex to avoid unrelated mutex contention issues. `_lastCount` has nothing to do with _streamMutex, so this is just asking for trouble. Each mutex should cover as small a number of cases/variables as possible to reduce contention andthe likelyhood of accidentaly trying to lock the mutex twice. (I think in the original code, _streamMutex was already locked for some other reason, so adding another mutex wouldn't help anything.)
 
However you may be able to get rid of the mutex entirely.

std::atomic<int> _lastCount{0}; // in the class header

auto count = ++_lastCount; // safe since _lastCount is atomic.
bool lastTaskDone = count >= _taskCount;
return lastTaskDone;

Reply:
Frankly, I'm a bit hesitant to do this now as this change means a possible shift in the locking behavior of these streaming classes compared with the current implementation. My intent here was not to change anything in the behavior of the SSI branch of the resulting streaming. I propose that we do this change (as per your suggestion above) as well as others that may modify the behavior of the streaming classes in a separate ticket. In this case, we would always see an effect of the changes.